### PR TITLE
Auth routes

### PR DESCRIPTION
**Pull request by Ofir-Smuha**
- Added authentication via routes

### CODE REVIEW

File: src/actions/user.actions.js [hunk 1]
```diff
@@ -38,7 +38,11 @@ export const getUserInfoWithToken = ({ token }: { token: string }) =>
     }
   });
 
-export const saveTokenToLocalStorage = ({ token }: { token: string }) => ({
+export const saveTokenToLocalStorage = ({
```
Comment (morsdyce) [inline src/actions/user.actions.js]: If you don't expect more than 1 parameter for an action then you can just have it as a single parameter (no destructuring).

I generally think it's fine to leave out destructuring until I have more than 2 parameters or if I have more than 1 parameter and one of them is a boolean.
since I don't want to see an action like this: myAction(token, true) - which makes me have no idea what true means without going to the action declaration.

File: src/components/app.js [hunk 2]
```diff
@@ -19,51 +20,88 @@ import NewIssue from 'components/issues/new-issue/NewIssue';
...
-  saveTokenToLocalStorage: () => void
+  saveTokenToLocalStorage: () => void,
+  isAuthenticated: string | null
 };
 
 type OwnProps = {};
 
 class App extends React.Component<ConnectedProps & OwnProps> {
   componentWillMount() {
     const user = get('user', loadFromStorage('auth'));
```
Comment (morsdyce) [inline src/components/app.js]: Loading directly from localStorage in a component is a bad practice, it is better handled in the redux layer using something like redux-persist.

So if the user has logged in and the reducer was updated that value will stay for the next page loads (and we won't need to load it directly ourselves).

Comment (morsdyce) [reply src/components/app.js]: In addition please do not use componentWillMount as it is deprecated and will be removed in the next react releases

File: src/components/app.js [hunk 3]
```diff
@@ -19,51 +20,88 @@ import NewIssue from 'components/issues/new-issue/NewIssue';
...
-    if (user) {
+
+    if (user && !isNull(user.token)) {
```
Comment (morsdyce) [inline src/components/app.js]: Since you only need the user token, there is no need to get the entire user object.
Instead of doing this check you can pull the token directly using the get function above and just check if you have a token

File: src/components/app.js [hunk 4]
```diff
@@ -19,51 +20,88 @@ import NewIssue from 'components/issues/new-issue/NewIssue';
...
       this.props.saveTokenToLocalStorage(user);
       this.props.getUserInfoWithToken(user);
+    } else {
+      const searchParams = get('search', this.props.location);
```
Comment (morsdyce) [inline src/components/app.js:43]: This is code that should be moved to the separate route we talked about which handles getting the token from github so this behaviour will be in 1 place and won't affect the entire application

File: src/components/common/PrivateRoute.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import { Route, Redirect } from 'react-router-dom';
+
+function PrivateRoute({ component: Component, isAuthenticated, ...rest }) {
+  return (
+    <Route
+      {...rest}
+      render={props =>
+        isAuthenticated ? (
```
Comment (morsdyce) [inline src/components/common/PrivateRoute.js]: This is very hard to read. Please avoid ternary statements when both outcomes are components (the only exception is when it's super small) in general it's better to avoid them.

It would be much easier to read as:
```
  if (!isAuthenticated) {
     return <Redirect to={{ pathname: '/login' }} />;
  }

  return <Component {...props} />;
```

File: src/reducers/user.reducer.js [hunk 6]
```diff
@@ -6,12 +6,11 @@ import {
   SET_AUTH_ERROR,
   RESET_AUTH_ERROR,
   SET_USER_IN_STATE,
-  FETCH_USER_REPOSITORIES,
   SET_USER_REPOSITORIES
 } from 'actions/user.actions';
 
 const initialState = {
-  token: null,
+  token: false,
```
Comment (morsdyce) [inline src/reducers/user.reducer.js]: why did you change the default value to false? it makes much more sense as null (which indicates you don't have one)

File: src/reducers/user.reducer.js [hunk 7]
```diff
@@ -29,11 +28,15 @@ export default handleActions(
     [SAVE_TOKEN_TO_LOCAL_STORAGE]: (
       state: UserState,
       { payload }: { payload: Object }
-    ) => set('token', payload.token, state),
+    ) => set('token', !payload.token ? null : payload.token, state),
```
Comment (morsdyce) [inline src/reducers/user.reducer.js]: I think the previous way was fine, there's no need to check what the value is.
If you want to enforce it will be specifically null instead of undefined you can use a default value in the action

**Review by morsdyce — CHANGES_REQUESTED** [review]

File: src/components/app.js (hunk 4 above)
Comment (Ofir-Smuha) [reply src/components/app.js:43]: Yes, i first builded the authentication route and the logic setup. Next PR it will be as it should.

File: src/reducers/user.reducer.js (hunk 6 above)
Comment (Ofir-Smuha) [reply src/reducers/user.reducer.js]: There is a play between null, false and true here, because i do not render the app until i get null or true so the PrivateRoute component won't redirect me to Login page before i get any value. Set it to null and you will never be able to access any page.

File: src/actions/user.actions.js [hunk 8]
```diff
@@ -38,14 +38,10 @@ export const getUserInfoWithToken = ({ token }: { token: string }) =>
     }
   });
 
-export const saveTokenToLocalStorage = ({
-  token
-}: {
-  token: string | null
-}) => ({
+export const saveTokenToLocalStorage = (user = { token: null }) => ({
```
Comment (morsdyce) [inline src/actions/user.actions.js]: This isn't what I meant, There is no value in the user object here. the only thing you care about is the token.
Which means you can pass the token as a single argument and the payload would be just `token` without `user` being mentioned at all.

This would be different if you wanted to persist the entire user object and the action would be called saveUserToLocalStorage

**Review by morsdyce — CHANGES_REQUESTED** [review]
Please also fix conflicts caused by merging the styling branch

File: src/actions/user.actions.js (hunk 8 above)
Comment (Ofir-Smuha) [reply src/actions/user.actions.js]: The problem is that this action is being called also from another action that fetches the token with the code from GitHub, so there is not much to do with the data structure (otherwise the other action data will fail)

Comment (morsdyce) [reply src/actions/user.actions.js]: This is still incorrect, it should take the token only

Comment (Ofir-Smuha) [reply src/actions/user.actions.js]: Fixed
